From 635539e1a9db0ee27f074db7cd86f7bfae42b3f3 Mon Sep 17 00:00:00 2001
From: Richard Gong <richard.gong@intel.com>
Date: Mon, 26 Apr 2021 11:01:30 -0500
Subject: [PATCH 34/41] HSD #18016072895-2: crypto: intel_fcs: add sessionID at
 PSGSIGMA_TEARDOWN

commit 24bf962a8befea56de4a0f4ceb41f2886d5211fa from
https://github.com/altera-opensource/linux-socfpga.git
branch is socfpga-5.4.114-lts

Add SDM sessionID in PSGSIGMA_TEARDOWN command for FPGA attestation.

Signed-off-by: Richard Gong <richard.gong@intel.com>
Integrated-by: Abhishek Paliwal <paliwal.abhishek@windriver.com>
---
 drivers/crypto/intel_fcs.c           | 16 ++++++++++++++++
 include/uapi/linux/intel_fcs-ioctl.h |  2 ++
 2 files changed, 18 insertions(+)

diff --git a/drivers/crypto/intel_fcs.c b/drivers/crypto/intel_fcs.c
index 5765fc5bceb9..595806eb968b 100644
--- a/drivers/crypto/intel_fcs.c
+++ b/drivers/crypto/intel_fcs.c
@@ -39,6 +39,8 @@
 #define MEASUREMENT_CMD_MAX_SZ	4092
 #define MEASUREMENT_RSP_MAX_SZ	4092
 
+#define SIGMA_SESSION_ID_ONE	0x1
+#define SIGMA_UNKNOWN_SESSION	0xffffffff
 
 #define FCS_REQUEST_TIMEOUT (msecs_to_jiffies(SVC_FCS_REQUEST_TIMEOUT_MS))
 #define FCS_COMPLETED_TIMEOUT (msecs_to_jiffies(SVC_COMPLETED_TIMEOUT_MS))
@@ -191,6 +193,7 @@ static long fcs_ioctl(struct file *file, unsigned int cmd,
 	const struct firmware *fw;
 	char filename[FILE_NAME_SIZE];
 	size_t tsz, rsz, datasz;
+	uint32_t sid;
 	void *s_buf;
 	void *d_buf;
 	void *ps_buf;
@@ -677,7 +680,20 @@ static long fcs_ioctl(struct file *file, unsigned int cmd,
 		break;
 
 	case INTEL_FCS_DEV_PSGSIGMA_TEARDOWN:
+		if (copy_from_user(data, (void __user *)arg, sizeof(*data))) {
+			dev_err(dev, "failure on copy_from_user\n");
+			return -EFAULT;
+		}
+
+		sid = data->com_paras.tdown.sid;
+		if ((sid != SIGMA_SESSION_ID_ONE) &&
+			(sid != SIGMA_UNKNOWN_SESSION)) {
+			dev_err(dev, "Invalid session ID:%d\n", sid);
+			return -EFAULT;
+		}
+
 		msg->command = COMMAND_FCS_PSGSIGMA_TEARDOWN;
+		msg->arg[0] = sid;
 		priv->client.receive_cb = fcs_vab_callback;
 		ret = fcs_request_service(priv, (void *)msg,
 					  FCS_REQUEST_TIMEOUT);
diff --git a/include/uapi/linux/intel_fcs-ioctl.h b/include/uapi/linux/intel_fcs-ioctl.h
index 40963b187f2b..24c3aeed6b91 100644
--- a/include/uapi/linux/intel_fcs-ioctl.h
+++ b/include/uapi/linux/intel_fcs-ioctl.h
@@ -117,9 +117,11 @@ struct fcs_random_number_gen {
 /**
  * struct fcs_psgsigma_teardown
  * @teardown
+ * @sid: the session ID
  */
 struct fcs_psgsigma_teardown {
 	bool teardown;
+	uint32_t sid;
 };
 
 /**
-- 
2.18.1

